src/Components/Aspire.OpenAI/AspireOpenAIClientBuilderEmbeddingGeneratorExtensions.cs

```diff
@@ -0,0 +1,111 @@
+    /// <returns>A <see cref="EmbeddingGeneratorBuilder{TInput, TEmbedding}"/> that can be used to build a pipeline around the inner <see cref="IEmbeddingGenerator{TInput, TEmbedding}"/>.</returns>
+    public static EmbeddingGeneratorBuilder<string, Embedding<float>> AddEmbeddingGenerator(
+        this AspireOpenAIClientBuilder builder,
+        string? deploymentName = null,
+        bool disableOpenTelemetry = false)
+    {
+        return builder.HostBuilder.Services.AddEmbeddingGenerator(
+            services => CreateInnerEmbeddingGenerator(services, builder, deploymentName, disableOpenTelemetry));
+    }
+
+    /// <summary>
+    /// Registers a keyed singleton <see cref="IEmbeddingGenerator{TInput, TEmbedding}"/> in the services provided by the <paramref name="builder"/>.
+    /// </summary>
+    /// <param name="builder">An <see cref="AspireOpenAIClientBuilder" />.</param>
+    /// <param name="serviceKey">The service key with which the <see cref="IEmbeddingGenerator{TInput, TEmbedding}"/> will be registered.</param>
+    /// <param name="deploymentName">Optionally specifies which model deployment to use. If not specified, a value will be taken from the connection string.</param>
+    /// <param name="disableOpenTelemetry">Optional. If <see langword="true"/>, skips registering open telemetry support in the <see cref="IEmbeddingGenerator{TInput, TEmbedding}"/> pipeline.</param>
+    /// <returns>A <see cref="EmbeddingGeneratorBuilder{TInput, TEmbedding}"/> that can be used to build a pipeline around the inner <see cref="IEmbeddingGenerator{TInput, TEmbedding}"/>.</returns>
+    public static EmbeddingGeneratorBuilder<string, Embedding<float>> AddKeyedEmbeddingGenerator(
+        this AspireOpenAIClientBuilder builder,
+        string serviceKey,
+        string? deploymentName = null,
+        bool disableOpenTelemetry = false)
+    {
+        return builder.HostBuilder.Services.AddKeyedEmbeddingGenerator(
+            serviceKey,
+            services => CreateInnerEmbeddingGenerator(services, builder, deploymentName, disableOpenTelemetry));
+    }
+
+    private static IEmbeddingGenerator<string, Embedding<float>> CreateInnerEmbeddingGenerator(
+        IServiceProvider services,
+        AspireOpenAIClientBuilder builder,
+        string? deploymentName,
+        bool disableOpenTelemetry)
+    {
+        var openAiClient = builder.ServiceKey is null
+            ? services.GetRequiredService<OpenAIClient>()
+            : services.GetRequiredKeyedService<OpenAIClient>(builder.ServiceKey);
+
+        deploymentName ??= GetRequiredDeploymentName(builder);
+        var result = openAiClient.AsEmbeddingGenerator(deploymentName);
+
+        return disableOpenTelemetry
+            ? result
+            : new OpenTelemetryEmbeddingGenerator<string, Embedding<float>>(result);
+    }
+
+    private static string GetRequiredDeploymentName(AspireOpenAIClientBuilder builder)
```

Comment:
Can we refactor this code so it isn't duplicated between these two builder classes?

Reply:
I can factor out a shared utility for it.

Reply:
Done